### PR TITLE
fix: persist virtual views when refreshing dbt

### PR DESCRIPTION
**Pull request by joaoviana**
Closes: [#11749](https://github.com/lightdash/lightdash/issues/11749)

### Description:

This PR allows the persistence of `virtual views`/`custom explores` when the user refreshes dbt.
It also moved these below all _default_ explores

Screenshot 2024-10-02 at 15 32 48

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/models/ProjectModel/ProjectModel.ts [hunk 1]
```diff
@@ -2042,23 +2050,13 @@ export class ProjectModel {
...
-        const toAddToCached: Explore = {
-            name,
-            tags: [],
-            label: friendlyName(name),
-            tables: translatedToExplore.tables,
-            baseTable: name,
-            groupLabel: 'Custom explores',
-            joinedTables: [],
-            targetDatabase: SupportedDbtAdapter.POSTGRES,
-        };
```
Comment (joaoviana) [inline packages/backend/src/models/ProjectModel/ProjectModel.ts:2061]: this was unnecessary - also wrong 😅

File: packages/backend/src/models/ProjectModel/ProjectModel.ts [hunk 2]
```diff
@@ -933,14 +933,22 @@ export class ProjectModel {
             'ProjectModel.saveExploresToCache',
             {},
             async () => {
+                // Get custom explores/virtual views before deleting them
+                const customExplores = await this.database(
+                    CachedExploreTableName,
+                )
+                    .select('explore')
+                    .where('project_uuid', projectUuid)
+                    .whereRaw("explore->>'type' = ?", [ExploreType.CUSTOM]);
```
Comment (rephus) [inline packages/backend/src/models/ProjectModel/ProjectModel.ts:942]: we'll need now a new ticket to allow users `delete` custom explores in the UI

**Review by rephus — APPROVED** [review]
This works great, 

There is only one weird thing if you have some filtered tables on settings. We should probably create a separate ticket for this. 

When creating the custom explore, you can see it on the explore, but it doesn't show up on the list

I think we should `add it to the list of selected explores` on custom explore creation so they appear automatically. Forcing custom explores to always appear on the list might not be ideal, specially since we can't remove them 

Screenshot from 2024-10-03 08-37-29
Screenshot from 2024-10-03 08-36-35
[Screencast from 03-10-24 08:36:56.webm](https://github.com/user-attachments/assets/dfc46e9a-3e0f-45c4-b678-16fbc52987ac)

**Comment by joaoviana** [comment]
No filtering anymore for custom explores/virtual views

Screenshot 2024-10-03 at 10 58 33

**before adding:**
Screenshot 2024-10-03 at 10 58 39

**after adding:**
Screenshot 2024-10-03 at 10 59 02

**Comment by rephus** [comment]
is the e2e error on catalog.cy.ts related ?

File: packages/common/src/types/explore.ts [hunk 3]
```diff
@@ -38,6 +38,11 @@ export type CompiledTable = TableBase & {
     uncompiledSqlWhere?: string;
 };
 
+export enum ExploreType {
+    CUSTOM = 'custom',
```
Comment (rephus) [inline packages/common/src/types/explore.ts:42]: should we use `virtual` here ?

Comment (joaoviana) [reply packages/common/src/types/explore.ts:42]: Agree, I should change! ty

**Comment by joaoviana** [comment]
> is the e2e error on catalog.cy.ts related ?
> 
> Is this because you manually changed the selected tables on render, and that broke some e2e tests ? If that's the case, we can merge

This was exactly why @rephus ! 
Reset the db to have a clean slate and all are passing 😄

**Pull request merged**